spec.md

```diff
@@ -1293,10 +1303,15 @@ The CO MUST implement the specified error recovery behavior when it encounters t
 
 Controller Plugin MUST implement this RPC call if it has `PUBLISH_UNPUBLISH_VOLUME` controller capability.
 This RPC is a reverse operation of `ControllerPublishVolume`.
-It MUST be called after all `NodeUnstageVolume` and `NodeUnpublishVolume` on the volume are called and succeed.
+It MUST be called after both `NodeUnstageVolume` and `NodeUnpublishVolume` on the volume are called and succeed unless the plugin has the `UNPUBLISH_FENCE` capability.
 The Plugin SHOULD perform the work that is necessary for making the volume ready to be consumed by a different node.
 The Plugin MUST NOT assume that this RPC will be executed on the node where the volume was previously used.
 
+If the plugin has the `UNPUBLISH_FENCE` capability, the CO MAY specify `fence` as `true`, in which case the SP MUST ensure that the node may no longer access the volume before returning a successful response.
+This results in a transition into one of the `QUARANTINE` states where the node must be cleaned up without being able to access the volume like usual.
+This is intended to cut off an unreachable node from accessing volumes so those volumes may be safely published to another node.
+Once in one of the `QUARANTINE` states the volume MAY NOT be published to that node again until appropriate cleanup has happened using `NodeUnpublishVolume` and `NodeUnstageVolume` (if applicable).
```

Comment:
Thinking from k8s point of view - how will we actually perform the cleanup on the node? Say a node was shutdown with volumes attached to it and hence CO calls `ControllerUnpublishVolume` with `fence: true`. The node comes back:

1. Should node have some taint to prevent scheduling of pods with these volumes?
2. We could implement something during volume reconstruction in kubelet.  But I am still unsure about how will kubelet mark those volumes as "clean".
3. Are we proposing additional state checks in kube-controller-manager that could detect if volume needs to be detached with `fenced: true`? Currently KCM won't detach volumes from unresponsive/shutdown nodes at all.

Reply:
@gnufied Can you help review this KEP?  https://github.com/kubernetes/enhancements/pull/1116

Reply:
Thanks. I did not know that k8s KEP was updated with CSI spec change. I will take a look.